src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5557,56 +5570,65 @@ public override void VisitForEachIterationVariables(BoundForEachStatement node)
                 };
 #pragma warning restore IDE0055 // Fix formatting
 
-            // declare and assign all iteration variables
-            foreach (var iterationVariable in node.IterationVariables)
+            if (!(node.DeconstructionOpt is null))
             {
-                var state = NullableFlowState.NotNull;
-                if (!sourceState.HasNullType)
-                {
-                    TypeWithAnnotations destinationType = iterationVariable.TypeWithAnnotations;
+                var assignment = node.DeconstructionOpt.DeconstructionAssignment;
 
-                    if (iterationVariable.IsRef)
+
+                // Visit the assignment as a deconstruction with an explicit type
+                VisitDeconstructionAssignmentOperator(assignment, sourceState);
+            }
+            else
+            {
+                foreach (var iterationVariable in node.IterationVariables)
+                {
+                    var state = NullableFlowState.NotNull;
+                    if (!sourceState.HasNullType)
                     {
-                        // foreach (ref DestinationType variable in collection)
-                        if (IsNullabilityMismatch(sourceType, destinationType))
+                        TypeWithAnnotations destinationType = iterationVariable.TypeWithAnnotations;
+
+                        if (iterationVariable.IsRef)
+                        {
+                            // foreach (ref DestinationType variable in collection)
+                            if (IsNullabilityMismatch(sourceType, destinationType))
+                            {
+                                var foreachSyntax = (ForEachStatementSyntax)node.Syntax;
+                                ReportNullabilityMismatchInAssignment(foreachSyntax.Type, sourceType, destinationType);
+                            }
+                            state = sourceState.State;
+                        }
+                        else
                         {
-                            var foreachSyntax = (ForEachStatementSyntax)node.Syntax;
-                            ReportNullabilityMismatchInAssignment(foreachSyntax.Type, sourceType, destinationType);
+                            // foreach (DestinationType variable in collection)
+                            // foreach (var variable in collection)
+                            // and asynchronous variants
+                            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                            Conversion conversion = node.ElementConversion.Kind == ConversionKind.UnsetConversionKind
+                                ? _conversions.ClassifyImplicitConversionFromType(sourceType.Type, destinationType.Type, ref useSiteDiagnostics)
+                                : node.ElementConversion;
```

Comment:
I'm not sure this is correct, because this conversion will be using types from initial binding. There's probably a nested generic substitution with a user-defined conversion scenario you could get into where this will have the wrong type arguments.

Reply:
Actually, this is probably fine.

---
In reply to: [280929577](https://github.com/dotnet/roslyn/pull/35445#discussion_r280929577) [](ancestors = 280929577)